### ISSUE TITLE
support for ssl (and therefore https) not working in Python 3

### ISSUE THREAD

**Issue by jackjansen** [webpy/webpy#471]
It seems SSL support is not working under Python 3. 

At first I thought this was a shallow error (the `from wsgiserver.ssl_pyopenssl import pyOpenSSLAdapter` in `httpserver.WSGIServer.create_ssl_adapter()` is treated as an absolute import in Python 3 and it should be relative), but then an undefined `CP_fileobject` showed up because wsgiserver3 has that object replaced by a `CP_makefile` function. At this point I no longer understand the code, so I'm afraid I cannot offer a fix.

**Comment by jackjansen** [comment]
I'm sorry, but I have since then moved away from web.py to Flask (unfortunately). What I do remember is that the core issue seems to be that web.py uses a slurped wsgiserver from CherryPy and the issues were in there. 

In cherrypy they've turned that server into a standalone project [cheroot](https://github.com/cherrypy/cheroot) so also switching web.py to that should probably fix the issue. I think I remember seeing a note from the cherrypy people that they separated out the wsgiserver specifically because various projects were using it without using the rest of cherrypy (like web.py).

**Comment by mowbraym** [comment]
I am using the sample code below. I've created the certificate files and the code works BUT on each access to the page it prompts *at the server* for the certificate passphrase (as shown in session below the code below). Any ideas to stop that happening?

```python
import web

urls = ("/.*", "hello")
app = web.application(urls, globals())

class hello:
    def GET(self):
        return 'Hello, world!'

from cheroot.server import HTTPServer
from cheroot.ssl.builtin import BuiltinSSLAdapter

HTTPServer.ssl_adapter = BuiltinSSLAdapter(
        certificate='./cert.pem',
        private_key='./key.pem')

if __name__ == "__main__":
    app.run()
```

```
 $ python3 cert.py 1234
Enter PEM pass phrase:
https://0.0.0.0:1234/
Enter PEM pass phrase:
192.168.1.10:60501 - - [30/Aug/2019 16:38:15] "HTTP/1.1 GET /" - 200 OK
192.168.1.10:60501 - - [30/Aug/2019 16:38:15] "HTTP/1.1 GET /favicon.ico" - 200 OK

```

**Comment by iredmail** [comment]
I'm afraid you need to remove the passphrase. :(

**Comment by mowbraym** [comment]
I tried that but OpenSSL seems to be forcing one on me ... I'll look for a way around that.

**Comment by iredmail** [comment]
if i remember correctly, years ago i read some post which mentions the webpy builtin web server is not recommended for production use, maybe i was wrong. But running Apache/Nginx/Lighttpd/... web server to handle ssl cert may be better. Just my 2 cents.

**Comment by mowbraym** [comment]
Yep I hear you - saw a lot of that myself, and removing the passphrase just seems to support that. For completeness, I generated a new certificate/key pair using:
`$ openssl req  -nodes -new -x509  -keyout key.pem -out cert.pem`
and it does now work without a passphrase, but at what cost to security I wonder?

**Comment by jackjansen** [comment]
@mowbraym if you're deploying in a container and/or using the certificate only for this one service you needn't worry overmuch about the cost to security. If you're deploying on a multiuser machine with lots of services running: yes, you should worry, because there are too many attack surfaces where an intruder could get access to your key and then impersonate many services. But in a modern deployment style (with a container per service) the attack surface is just your service.
